### PR TITLE
FSE: Disable Block Directory on Simple sites

### LINKED ISSUE #43677: Gutenberg 8.4.0: Disable block library
The Block Library is incompatible with Simple sites and will always display this message when searching:

> No blocks found in your library.
> Please contact your site administrator to install new blocks.

Disable it for Simple sites. See https://github.com/WordPress/gutenberg/pull/23389.

### PR DESCRIPTION
**Pull request by WunderBart**
#### Changes proposed in this Pull Request

Disable Block Directory via FSE's Block Inserter Modifications module by unregistering the `block-directory` plugin and registering our own to render `no results` state.

⚠️ This implementation disables Block Inserter on Simple sites **only** (`wp-admin` + `iframed` editors).
⚠️ This [module](https://github.com/Automattic/wp-calypso/tree/e9e41bdda49da4e93f16580f9fdb1acdcb1de240/apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications/disable-block-directory) should be removed once we start supporting Block Directory.

#### Testing instructions

- Enable edge sticker on your Simple test site,
- Open that site sandboxed and verify that it's using Gutenberg v8.5.1,
- Run `yarn dev --sync` to upload local FSE `dev` build to your sandbox,
- Once FSE has uploaded, refresh the page and open the Block Inserter,
- Try searching for something that should return no results - you should see the `No results found.` message,
- Try searching for `css` - you should see the no results message **and** the Contextual Tip about our Customizer,
- Build and upload the plugin to your AT site and make sure the Block Directory isn't disabled there.

#### Screenshots

| Before<br>(no results)  | Before<br>(no res. + cont. tip) | After<br>(no results) | After<br>(no res. + cont. tip) |
| ------------- | ------------- | ------------- |  ------------- |
| image | image | image | image |

Fixes #43677

### CODE REVIEW

File: apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications/disable-block-directory/index.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	registerPlugin( 'disable-block-registry', {
+		render: () => (
+			<__experimentalInserterMenuExtension>
+				{ ( { hasResults } ) => {
+					if ( ! hasResults ) {
+						// We need to define our own 'no results' until we enable Block Directory.
+						return (
+							<div className="disable-block-directory__no-results">
+								{ /* translators: Displayed when block search returns no results. */ }
+								{ __( 'No results found.', 'full-site-editing' ) }
```
Comment (a8ci18n) [inline apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications/disable-block-directory/index.js:35]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **58** times:
`translate( 'No results found.', { context: 'search results'} )` *ES Score: 11*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/disable-block-directory)

File: apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications/disable-block-directory/index.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		registerPlugin( 'disable-block-registry', {
+			render: () => (
+				<__experimentalInserterMenuExtension>
+					{ ( { hasResults } ) => {
+						if ( ! hasResults ) {
+							// We need to define our own 'no results' until we enable Block Directory.
+							return (
+								<div className="disable-block-directory__no-results">
+									{ /* translators: Displayed when block search returns no results. */ }
+									{ __( 'No results found.', 'full-site-editing' ) }
```
Comment (a8ci18n) [inline apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications/disable-block-directory/index.js:37]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **58** times:
`translate( 'No results found.', { context: 'search results'} )` *ES Score: 11*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/disable-block-directory)

**Comment by fullofcaffeine** [comment]
Hi Bart :wave:! 

I've tested this changeset in a test site of mine, here are the steps I followed:

1. I made sure `fullofcaffeine.wordpress.com` was sandboxed;
2. I `shh`'ed into my sandbox, `cd`'ed into `public_html` and ran `svn reset -R .` then `svn st | grep '^?' | awk '{print }' | xargs rm -rf` to make sure to remove any untracked leftovers;
3. I then came back to my local box, `cd`ed to `wp-calypso` and checked out your branch, and `cd`ed to the FSE sub-directory to run `yarn dev --dev`;
4. Accessed `https://fullofcaffeine.wordpress.com/wp-admin/post-new.php?post_type=page`/. Made sure I was using Gutenberg `8.5.1`:

version


5. Brought up the contextual picker on the page, by clicking the rounded square + button. Searched for an unexistent block, got this:

unexistingblock

6. Tried the same thing, but this time used the button on the top-left corner of the content area, in the main toolbar (the picker that appears in the sidebar). Got this:

unexisting2

7. Then proceeded to try searching for `css`, first with the page inline picker:

fr1

8. Then with the sidebar picker:

fr2


So, it works fine when the sidebar picker is used but doesn't for the inline picker. Perhaps I'm missing something?

I'm yet to test the AT site scenario but will do so shortly and post my findings here.

**Comment by fullofcaffeine** [comment]
> Build and upload the plugin to your AT site and make sure the Block Directory isn't disabled there.

1. I created a new ephemeral AT site and `scp`'ed the `full-site-editing-plugin` sub-dir after building the whole project with `yarn build`. Made sure it was activated;
2. I also installed Gutenberg `8.5.1` and activated it;
3. Brought up the picker from the page and tried searching for an unexistent block. Got a "No results found." message:



4. Brought up the sidebar picker and searched for an unexistent block. Got a "No blocks found in your library", which I think is the expected result?



I noticed that the "page" inline picker also has a slightly different UI and behavior here as well. Is that expected?

File: apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications/disable-block-directory/index.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			render: () => (
+				<__experimentalInserterMenuExtension>
+					{ ( { hasResults } ) => {
+						if ( ! hasResults ) {
+							// We need to define our own 'no results' until we enable Block Directory.
+							return (
+								<div className="disable-block-directory__no-results">
+									<Icon icon={ blockDefault } />
+									{ /* translators: Displayed when block search returns no results. */ }
+									<p>{ __( 'No results found.', 'full-site-editing' ) }</p>
```
Comment (a8ci18n) [inline apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications/disable-block-directory/index.js:39]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **58** times:
`translate( 'No results found.', { context: 'search results'} )` *ES Score: 11*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/disable-block-directory)

**Comment by WunderBart** [comment]
Thanks for the thorough review, @fullofcaffeine 🙇 

> 7. Then proceeded to try searching for css, first with the page inline picker. Expected to see the contextual tip, but it didn't appear:

Nice catch! This seems to be an issue with the  `>=8.3.0`, where the [Quick Inserter](https://github.com/WordPress/gutenberg/blob/1bec21feb1ba8110fdd11b01c84dc2a3849884a1/packages/block-editor/src/components/inserter/quick-inserter.js) is not using the `__experimentalInserterMenuExtension` slot that we need for [rendering Contextual Tips](https://github.com/Automattic/wp-calypso/blob/837afc0e09524316c292771c3d244717a3f1d44b/apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications/contextual-tips.js#L25-L26) anymore. @retrofox, is that something you'd want to look into?

> (On AT site) I noticed that the "page" inline picker also has a slightly different UI and behavior here as well. Is that expected?

Yep, this is how it looks without the disabling script on - AT site should not be affected at all by it and, according to your review, it's not 👍 

---

I added the block icon to the no results page for more consistency between inserters (+ updated screenshots in the description).

**Comment by sirreal** [comment]
> ⚠️ This module should be removed once we start supporting Block Directory.

I don't expect to remove this, Simple sites won't support one-click block installs but will offer an upgrade path that would probably be implemented on top of the solution we're working on here.

**Review by sirreal — CHANGES_REQUESTED** [review]
I'm not sure what's going on, but the notice is either rendered too eagerly or needs to be cleaned up when there are results:

Screen Shot 2020-07-14 at 16 32 09

**Comment by sirreal** [comment]
If we're only interested in shipping this functionality to Simple sites, it's probably best to implement it as a WPCOM diff. We don't need to get into all the FSE complexity, it may be as simple as dequeueing a script.

**Comment by sirreal** [comment]
Let's go with D46418-code instead of working through FSE.

**Pull request closed without merging**